worker/src/backgroundMigrations/backgroundMigrationManager.ts

```diff
@@ -0,0 +1,178 @@
+import { randomUUID } from "crypto";
+import { IBackgroundMigration } from "./IBackgroundMigration";
+import { prisma, Prisma } from "@langfuse/shared/src/db";
+import { instrumentAsync, logger } from "@langfuse/shared/src/server";
+
+export class BackgroundMigrationManager {
+  private static workerId = randomUUID();
+  private static activeMigration:
+    | {
+        id: string;
+        name: string;
+        args: Record<string, unknown>;
+        migration: IBackgroundMigration;
+      }
+    | undefined;
+
+  private static async heartBeat(): Promise<void> {
+    if (!BackgroundMigrationManager.activeMigration) {
+      return;
+    }
+
+    await prisma.backgroundMigration.updateMany({
+      where: {
+        id: BackgroundMigrationManager.activeMigration.id,
+        workerId: BackgroundMigrationManager.workerId,
+        finishedAt: null,
+        failedAt: null,
+      },
+      data: {
+        lockedAt: new Date(),
+      },
+    });
+
+    // Schedule next heartbeat in 15s
+    setTimeout(BackgroundMigrationManager.heartBeat, 15 * 1000);
+  }
+
+  public static async run(): Promise<void> {
+    await instrumentAsync({ name: "background-migration-run" }, async () => {
+      let migrationToRun = true;
+
+      while (migrationToRun) {
+        await prisma.$transaction(
+          async (tx) => {
+            // Read background migrations from database
+            const migration = await tx.backgroundMigration.findFirst({
+              where: {
+                finishedAt: null,
+                failedAt: null,
+              },
+              orderBy: { name: "asc" },
+            });
+
+            // Abort if there is no migration to run or migration was locked less than 60s ago
+            // We do not check lockedAt in the DB query, because findFirst might return other uncompleted migrations
+            // which would lead to concurrent execution.
+            if (
+              !migration ||
+              (migration.lockedAt &&
+                migration.lockedAt > new Date(Date.now() - 60 * 1000))
+            ) {
+              logger.info("No background migrations to run");
+              migrationToRun = false;
+              return;
+            }
+
+            logger.info(`Found background migrations ${migration.name} to run`);
+
+            // Acquire lock
+            await tx.backgroundMigration.update({
+              where: {
+                id: migration.id,
+              },
+              data: {
+                workerId: BackgroundMigrationManager.workerId,
+                lockedAt: new Date(),
+              },
+            });
```

Comment:
thought we could also add the migration id to the otel span

Reply:
@maxdeichmann IMO that wouldn't be very useful as all of them are a one-time thing. We have the wrapper span for all migrations that we perform and if we link it to logs that should be sufficient.

Reply:
ok makes sense